src/Lecture2.hs

```diff
@@ -159,8 +178,92 @@ data Knight = Knight
     , knightAttack    :: Int
     , knightEndurance :: Int
     }
+    deriving (Show)
+
+data DragonType = RedDragon | BlackDragon | GreenDragon deriving (Show)
+
+
+data Dragon = Dragon
+    { dragonHealth :: Int
+    , dragonFirePower :: Int
+    , dragonType :: DragonType
+    }
+    deriving (Show)
+    
+{- I bit confused with this part ( -}
+data Chest treasure = Chest
+    { chestGold     :: Int
+    , chestTreasure :: treasure
+    }
+    deriving (Show)
```

Comment:
That's the correct type here 👍🏻 To model the problem in the desired way, you need to change the definition o the `Dragon` type to capture the described invariant. So something like this should work:

```haskell
data DragonInfo = DragonInfo
    { dragonInfoHealth :: Int
    , dragonInfoFirePower :: Int
    }

data Dragon a
    = Green DragonInfo Gold
    | Red DragonInfo (Chest a)
    | Black DragonInfo (Chest a)
```